### PR TITLE
Observability improvement

### PR DESCRIPTION
**Pull request by msftcoderdjw**
1. Add more metrics.
2. Integrate otel-metrics in Symphony-api and Symphony-k8s.
3. Refactor the logger interface to allow pass context. (Will decorate the log entry with diagnostics context/activity context. Diagnostics context includes the correlationid, resourceid, traceid, spanid. This helps developers to correlate the logs with each other and developers won't need to explicitly inject traceid/spanid when writing logs.)
5. Print caller info in zap logger and logrus. 
6. Change the log format to json which is better for otel-integration.
7. Produce user diagnostics and user audits logs for external system (e.g. Azure Cloud.)
8. Advanced log and metrics experience with otel-ELK, otel-prometheus.

https://github.com/msftcoderdjw/eclipse-symphony/blob/users/jiadu/otel2/docs/symphony-book/observability/logs.md
https://github.com/msftcoderdjw/eclipse-symphony/blob/users/jiadu/otel2/docs/symphony-book/observability/metrics.md

Observability guidance for Symphony developers:
1. Always use **xxxCtx** (e.g. `InfofCtx`) and pass the context in logs. 
2. When adding new target / stage providers, keep in mind add user audit logs, user diagnostics logs and metrics for external system (e.g. Azure Cloud.)

### CODE REVIEW

File: coa/pkg/logger/coa-logger.go [hunk 1]
```diff
@@ -7,32 +7,47 @@
...
+	"runtime"
+	"sync"
 	"time"
 
+	"github.com/eclipse-symphony/symphony/coa/pkg/logger/hooks"
 	"github.com/sirupsen/logrus"
 )
 
 // daprLogger is the implemention for logrus
 type daprLogger struct {
```
Comment (Haishi2016) [inline coa/pkg/logger/coa-logger.go:22]: This is long overdue but we should rename this to coaLogger

File: coa/pkg/logger/coa-logger.go [hunk 2]
```diff
@@ -7,32 +7,47 @@
...
+
+	// sharedFieldsLock is the mutex for sharedFields
+	sharedFieldsLock sync.Mutex
+	// sharedFields is the fields that are shared among loggers
+	sharedFields logrus.Fields
+
+	callerSkip uint
 }
 
 var DaprVersion string = "unknown"
```
Comment (Haishi2016) [inline coa/pkg/logger/coa-logger.go:36]: We need to get rid of Dapr references

File: coa/pkg/logger/contexts/activitylog-context.go [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/valyala/fasthttp"
+)
+
+const (
+	ACTIVITY_HTTP_HEADER_PREFIX string = "X-Activity-"
```
Comment (Haishi2016) [inline coa/pkg/logger/contexts/activitylog-context.go:14]: This const uses a different casing

File: coa/pkg/apis/v1alpha2/bindings/http/metrics/metrics.go [hunk 4]
```diff
@@ -15,14 +15,14 @@ import (
 
 // Metrics is a metrics tracker for an api operation.
 type Metrics struct {
-	apiOperationLatency observability.Histogram
+	apiOperationLatency observability.Gauge
```
Comment (Haishi2016) [inline coa/pkg/apis/v1alpha2/bindings/http/metrics/metrics.go:18]: Same as previous comment on using Histogram - is there a reason to switch to gauge?

File: coa/pkg/logger/contexts/activitylog-context.go [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+	Activity_ResourceCloudId    string = "resourceId"
+	Activity_OperationName      string = "operationName"
+	Activity_Location           string = "location"
+	Activity_Category           string = "category"
+	Activity_CorrelationId      string = "correlationId"
+	Activity_Properties         string = "properties"
+
+	Activity_Props_CallerId      string = "caller-id"
```
Comment (Haishi2016) [inline coa/pkg/logger/contexts/activitylog-context.go:22]: Do we need both Activity_Props_CallerId and OTEL_Activity_Props_CallerId

File: api/pkg/apis/v1alpha1/managers/jobs/jobs-manager.go [hunk 6]
```diff
@@ -505,8 +514,8 @@ func (s *JobsManager) HandleJobEvent(ctx context.Context, event v1alpha2.Event)
 				return v1alpha2.NewCOAError(nil, "unsupported action", v1alpha2.BadRequest)
 			}
 		case "deployment":
-			log.Infof(" M (Job): handling deployment job %s, action: %s", job.Id, job.Action)
-			log.Infof(" M (Job): deployment spec: %s", string(job.Data))
+			log.InfofCtx(ctx, " M (Job): handling deployment job %s, action: %s", job.Id, job.Action)
+			log.InfofCtx(ctx, " M (Job): deployment spec: %s", string(job.Data))
```
[the remark this answers is not in the document]
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/managers/jobs/jobs-manager.go:518]: We won't log the entire context object. The context-hooks only retrieve the information related to diagnostics and user activities the append to log entry.

https://github.com/msftcoderdjw/eclipse-symphony/blob/users/jiadu/otel2/coa/pkg/logger/hooks/context-hook.go#L70

We can also filter what we want to log in central place in context-hooks.

File: api/pkg/apis/v1alpha1/providers/metrics/metrics.go [hunk 7]
```diff
@@ -30,23 +30,31 @@ const (
...
+	UpdateOperationType   string = "Update"
+	DeleteOperationType   string = "Delete"
+	ValidateOperationType string = "Validate"
+	RunOperationType      string = "Run"
 )
 
 // Metrics is a metrics tracker for a provider operation.
 type Metrics struct {
-	providerOperationLatency observability.Histogram
+	providerOperationLatency observability.Gauge
```
[the remark this answers is not in the document]
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/providers/metrics/metrics.go:50]: I agree. This is due to some external collectors such as Geneva doesn't well support otel-histogram. If we use histogram to emit the data, the data in Geneva is wrong. We have reported this issue to ME team. But it still needs several months to consume the release. Before that, let's use gauge at first. And I can log an issue to switch it back.

File: coa/pkg/apis/v1alpha2/bindings/http/metrics/metrics.go (hunk 4 above)
Comment (msftcoderdjw) [reply coa/pkg/apis/v1alpha2/bindings/http/metrics/metrics.go:18]: This is due to some external collectors such as Geneva doesn't well support otel-histogram. If we use histogram to emit the data, the data in Geneva is wrong. We have reported this issue to ME team. But it still needs several months to consume the release. Before that, let's use gauge at first.

File: coa/pkg/logger/coa-logger.go (hunk 2 above)
Comment (msftcoderdjw) [reply coa/pkg/logger/coa-logger.go:36]: Switch to use Coa as the prefix.

File: coa/pkg/logger/coa-logger.go (hunk 1 above)
Comment (msftcoderdjw) [reply coa/pkg/logger/coa-logger.go:22]: Done.

File: coa/pkg/logger/contexts/activitylog-context.go (hunk 3 above)
Comment (msftcoderdjw) [reply coa/pkg/logger/contexts/activitylog-context.go:14]: Switch to use Activity_HttpHeaderPrefix.

File: coa/pkg/logger/contexts/activitylog-context.go (hunk 5 above)
Comment (msftcoderdjw) [reply coa/pkg/logger/contexts/activitylog-context.go:22]: The intention is we can allow different key in logrus and when sending to otel-collector. So far, these are the same.

File: api/pkg/apis/v1alpha1/managers/jobs/jobs-manager.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/managers/jobs/jobs-manager.go:518]: Done.

File: api/pkg/apis/v1alpha1/managers/solution/solution-manager.go [hunk 8]
```diff
@@ -326,15 +328,15 @@ func (s *SolutionManager) Reconcile(ctx context.Context, deployment model.Deploy
...
 	if err != nil {
 		summary.SummaryMessage = "failed to plan for deployment: " + err.Error()
-		log.Errorf(" M (Solution): failed to plan for deployment: %+v", err)
+		log.ErrorfCtx(ctx, " M (Solution): failed to plan for deployment: %+v", err)
 		return summary, err
 	}
 
 	planBytes, _ := json.Marshal(plan)
-	log.Debugf(" M (Solution): deployment plan: %s", string(planBytes))
+	log.DebugfCtx(ctx, " M (Solution): deployment plan: %s", string(planBytes))
```
[the remark this answers is not in the document]
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/managers/solution/solution-manager.go:336]: Done.

File: api/pkg/apis/v1alpha1/managers/stage/stage-manager.go [hunk 9]
```diff
@@ -506,7 +509,7 @@ func (s *StageManager) HandleTriggerEvent(ctx context.Context, campaign model.Ca
 			}
 		}
 
-		log.Debugf(" M (Stage): HandleTriggerEvent after evaluation inputs 2: %v", inputs)
+		log.DebugfCtx(ctx, " M (Stage): HandleTriggerEvent after evaluation inputs 2: %v", inputs)
```
[the remark this answers is not in the document]
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/managers/stage/stage-manager.go:512]: Done.

File: api/pkg/apis/v1alpha1/managers/stage/stage-manager.go [hunk 10]
```diff
@@ -471,7 +474,7 @@ func (s *StageManager) HandleTriggerEvent(ctx context.Context, campaign model.Ca
 			}
 		}
 
-		log.Debugf(" M (Stage): HandleTriggerEvent before evaluation inputs 2: %v", inputs)
+		log.DebugfCtx(ctx, " M (Stage): HandleTriggerEvent before evaluation inputs 2: %v", inputs)
```
[the remark this answers is not in the document]
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/managers/stage/stage-manager.go:477]: Done.

File: api/pkg/apis/v1alpha1/managers/stage/stage-manager.go [hunk 11]
```diff
@@ -423,7 +426,7 @@ func (s *StageManager) HandleTriggerEvent(ctx context.Context, campaign model.Ca
 
 			eCtx := s.VendorContext.EvaluationContext.Clone()
 			eCtx.Inputs = triggerData.Inputs
-			log.Debugf(" M (Stage): HandleTriggerEvent evaluation inputs 1: %v", eCtx.Inputs)
+			log.DebugfCtx(ctx, " M (Stage): HandleTriggerEvent evaluation inputs 1: %v", eCtx.Inputs)
```
[the remark this answers is not in the document]
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/managers/stage/stage-manager.go:429]: Done.

File: api/pkg/apis/v1alpha1/providers/target/k8s/k8s.go [hunk 12]
```diff
@@ -372,34 +391,38 @@ func (i *K8sTargetProvider) removeService(ctx context.Context, namespace string,
...
-	log.Infof("  P (K8s Target Provider): removeDeployment namespace - %s, name - %s", namespace, name)
+	defer observ_utils.EmitUserDiagnosticsLogs(ctx, &err)
+	log.InfofCtx(ctx, "  P (K8s Target Provider): removeDeployment namespace - %s, name - %s", namespace, name)
 
 	if namespace == "" {
 		namespace = "default"
 	}
 
 	foregroundDeletion := metav1.DeletePropagationForeground
+	logger.GetUserAuditsLogger().InfofCtx(ctx, "  P (K8s Target Provider): Starting remove deployment under namespace - %s, name - %s", namespace, name)
```
[the remark this answers is not in the document]
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/providers/target/k8s/k8s.go]: Good suggestions. Renamed and moved.

File: api/pkg/apis/v1alpha1/managers/instances/instances-manager.go [hunk 13]
```diff
@@ -47,6 +47,7 @@ func (t *InstancesManager) DeleteState(ctx context.Context, name string, namespa
 	})
 	var err error = nil
 	defer observ_utils.CloseSpanWithError(span, &err)
+	defer observ_utils.EmitUserDiagnosticsLogs(ctx, &err)
```
Comment (LeiQL) [inline api/pkg/apis/v1alpha1/managers/instances/instances-manager.go:50]: Just curious, do we care about the order of logs emit and close span?

File: packages/helm/symphony/values.yaml [hunk 14]
```diff
@@ -17,6 +17,12 @@ observability:
   tracing:
     exporter:
       zipkin: true
+  otelCollector:
```
Comment (RemindD) [inline packages/helm/symphony/values.yaml:20]: Could you add comments for the meaning of newly added parameters?

File: api/pkg/apis/v1alpha1/managers/instances/instances-manager.go (hunk 13 above)
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/managers/instances/instances-manager.go:50]: No, I am not realize there's any case we need to enforce the order?

File: packages/helm/symphony/values.yaml (hunk 14 above)
Comment (msftcoderdjw) [reply packages/helm/symphony/values.yaml:20]: Good point. I can add some comments here.

**Comment by msftcoderdjw** [comment]
> I have a general question. How do we judge whether to emit user audit log, user diagnostic log and operation log?

It depends on your use-case. 
Both user audit logs and user diagnostics logs are generated to user collect endpoints. 
- In existing case, we will log the external system change (e.g. creating namespace in k8s target provider) as user audit logs and log the caller-id if the external system such as Azure Cloud provides this. 
- User diagnostics logs are recommended to be as a supplementary of symphony object status. Usually if the symphony object operations timed out, users can still diagnose the user diagnostics logs to understand the internal logs. Especially, if external system such as Azure Cloud provides the cloud resource id, we will also link the user diagnostics logs with the cloud resource id.

And for operation log, I guess you mean the application runtime logs in this table - https://github.com/msftcoderdjw/eclipse-symphony/blob/users/jiadu/otel2/docs/symphony-book/observability/logs.md#analyze-symphony-logs. It is the same as original console logs. But usually the application runtime logs are collected and seen by symphony service host people. We shouldn't log any PII or sensitive information in application runtime logs. And always remember to include context when printing application runtime logs since the context will include the trace/or correlation information to correlate all logs per request.

**Review by LeiQL — APPROVED** [review]

**Review by iwangjintian — APPROVED** [review]

**Review by RemindD — APPROVED** [review]

File: docs/symphony-book/observability/logs.md [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+    {{- end }}
+  ]
+}
+```
+
+Because application runtime logs are emitted by logrus, zaplogger or event fmt.print method, to collect as much logs as possible, to integrate application runtime logs with OTLP, we need to launch open telemetry collector with file receiver.
+
+Sample: [symphony-otel-forwarder.yaml](../../../packages/helm/symphony/templates/symphony-core/symphony-otelforwarder.yaml)
+
+## Deplay Symphony with advanced log experience
```
[the remark this answers is not in the document]
Comment (msftcoderdjw) [reply docs/symphony-book/observability/logs.md:217]: Fixed.

**Review by Haishi2016 — APPROVED** [review]

**Pull request merged**